les/utils/limiter.go

```diff
@@ -0,0 +1,395 @@
+	var f float64
+	if relWeight <= 0.001 {
+		f = 1
+	} else {
+		f = 0.001 / relWeight
+	}
+	f *= 1000000000
+	flatWeight, valueWeight = uint64(f), uint64(f*value)
+	if flatWeight == 0 {
+		flatWeight = 1
+	}
+	return
+}
+
+// Add adds a new request to the node queue belonging to the given id. Value belongs
+// to the requesting node. A higher value gives the request a higher chance of being
+// served quickly in case of heavy load or a DDoS attack. weight is a rough estimate
+// of the serving cost of the request. A lower weight also gives the request a
+// better chance.
+func (l *Limiter) Add(id enode.ID, address string, value float64, reqWeight uint) chan chan struct{} {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	process := make(chan chan struct{}, 1)
+	if l.quit {
+		close(process)
+		return process
+	}
+	if value > l.maxValue {
+		l.maxValue = value
+	}
+	if value > 0 {
+		// normalize value to <= 1
+		value /= l.maxValue
+	}
+	if reqWeight == 0 {
+		reqWeight = 1
+	}
+	if reqWeight > l.maxWeight {
+		l.maxWeight = reqWeight
+	}
+
+	if nq, ok := l.nodes[id]; ok {
+		nq.queue = append(nq.queue, request{process, reqWeight})
+		nq.sumWeight += reqWeight
+		nq.value = value
+		if address != nq.address {
+			// known id sending request from a new address, move to different address group
+			l.removeFromGroup(nq)
+			l.addToGroup(nq, address)
+		}
+	} else {
+		nq := &nodeQueue{
+			queue:      []request{{process, reqWeight}},
+			id:         id,
+			value:      value,
+			nextWeight: reqWeight,
+			sumWeight:  reqWeight,
+			groupIndex: -1,
+		}
+		nq.flatWeight, nq.valueWeight = selectionWeights(float64(reqWeight)/float64(l.maxWeight), value)
+		if len(l.nodes) == 0 {
+			l.cond.Signal()
+		}
+		l.nodes[id] = nq
+		if nq.valueWeight != 0 {
+			l.valueSelect.Update(nq)
+		}
+		l.addToGroup(nq, address)
+	}
+	l.sumWeight += reqWeight
+	if l.sumWeight > l.sumWeightLimit {
+		l.dropRequests()
+	}
+	return process
+}
+
+// update updates the selection weights of the node queue
+func (l *Limiter) update(nq *nodeQueue) {
+	flatWeight, valueWeight := selectionWeights(float64(nq.nextWeight)/float64(l.maxWeight), nq.value)
+	ag := l.addresses[nq.address]
+	ag.update(nq, flatWeight)
+	l.addressSelect.Update(ag)
+	if valueWeight != 0 {
+		nq.valueWeight = valueWeight
+		l.valueSelect.Update(nq)
+	}
+}
+
+// addToGroup adds the node queue to the given address group. The group is created if
+// it does not exist yet.
+func (l *Limiter) addToGroup(nq *nodeQueue, address string) {
+	nq.address = address
+	ag := l.addresses[address]
+	if ag == nil {
+		ag = &addressGroup{nodeSelect: NewWeightedRandomSelect(flatWeight)}
+		l.addresses[address] = ag
+	}
+	ag.add(nq)
+	l.addressSelect.Update(ag)
+}
+
+// removeFromGroup removes the node queue from its address group
+func (l *Limiter) removeFromGroup(nq *nodeQueue) {
+	ag := l.addresses[nq.address]
+	ag.remove(nq)
+	if len(ag.nodes) == 0 {
+		delete(l.addresses, nq.address)
+	}
+	l.addressSelect.Update(ag)
```

Comment:
If the address group is empty, it's equivalent that:
- `l.addressSelect.Update(ag)`
- `l.addressSelect.Remove(ag)`
right? Because the weight is 0.

Reply:
Perhaps it's better to use `Remove` here?

Reply:
I don't see your point here. `Update` is outside the `if` clause so the weight is either 0 or not. `Update` works fine for both cases. Why should we care about calling `Remove` explicitly in the empty case when the selector handles this already?

Reply:
yes, my mistake